src/librustc_typeck/collect.rs

```diff
@@ -1532,6 +1504,11 @@ fn type_scheme_of_item<'a,'tcx>(ccx: &CrateCtxt<'a,'tcx>,
                                 it: &hir::Item)
                                 -> ty::TypeScheme<'tcx>
 {
+    // Computing the type scheme of an item is a discrete task:
+    let item_def_id = ccx.tcx.map.local_def_id(it.id);
+    let _task = ccx.tcx.dep_graph.in_task(DepNode::TypeScheme(item_def_id));
+    ccx.tcx.dep_graph.read(DepNode::Hir(item_def_id)); // we have access to `it`
```

Comment:
So, one other way to structure this would be not have separate TypeScheme and TCache nodes at all, because essentially they describe the same thing. If the computation of `foo(Bar)`'s TypeScheme requires the TypeScheme of `Bar`, it would go to the `tcache` and try to look it up. If it's not there, push the `TypeScheme(Bar)` task (but without automatically adding an edge from task to subtask), compute and add it to the map, and then accessing the map would introduce a read-edge from `TypeScheme(foo(Bar))` to `TypeScheme(Bar)`. That would avoid adding artificial edges and lead to a cleaner, more stable dependency graph (in this case at least). More stable because the graph would not depend on (re-)execution order.
Correct me if I'm wrong.